### ISSUE TITLE
Support question:  border() function

### ISSUE THREAD

**Issue by cbailiss** [davidgohel/flextable#322]
Thanks for a great package.

My use case is as follows:
I have generated a table with flextable().
I wish to highlight a specific cell in the table by changing the background colour (easy, no problem here) and changing the border colours on all four sides (not so easy/sometimes impossible with documented functions).

The documented border functions hline() and vline() can be used, but to set the borders on one cell means I must call each function specifying the cell row/column number and the row/column number minus 1 (since these functions only set the border on one side of each cell).

Sometimes it is not possible at all using the documented functions, e.g. if the cell is in the first column, then it is not possible to set the left border of the one specific cell since vline() only sets the right border.
 
The function I am really looking for is a simple function where I can specify the cell row/column number and the individual border settings for each side of the cell.  I see this function already exists - called simply border() in:
https://github.com/davidgohel/flextable/blob/master/R/borders.R

However, the border() function is not documented - it is a "hidden" internal function.  Since the code I am writing will be part of a package submitted to CRAN, using a "hidden" function that may be changed seems a bad idea.

I see the note in the source code that this is an internal function.  So, my request is, please could this function be made public/supported?

I understand the risk of overlapping/conflicting settings, e.g. calling the function on two cells next to each other with different border settings.  However, I think currently it is too difficult (or impossible in some cases, as noted above) to set borders on specific cells - and if possible it would be better to make the border() function public but note in the documentation that care is needed when using on cells that are next to each other.

**Comment by cbailiss** [comment]
Using style() and fp_border() looks like it helps to set border styles on individual cells - including cells in the first column.  Is there a way to specify the cell side (left,right,top,bottom) with fp_border?
Using the undocumented border() still looks like the simplest method that gives full control over the borders of a single specific cell - but please tell me if I have missed/misunderstood something!

**Comment by davidgohel** [comment]
No, I don't want to, I've implemented other solutions, that make things easier for users and me ;)

See https://ardata-fr.github.io/flextable-book/format.html#borders

For example, border_outer(), border_inner_h() and border_inner_v().

All the following functions also support the row and column selector `i` and `j`. 

hline(): set bottom borders (inner horizontal)

vline(): set right borders (inner vertical)

hline_top(): set the top border (outer horizontal)

hline_bottom(): set the bottom border (outer horizontal)

vline_left(): set the left border (outer vertical)

vline_right(): set the right border (outer vertical)

That said, if as you want to use `border()` knowing you must be careful with border overlapping, you can. The meaning of *internal* (from my understanding) is it is for package developer or advanced users. The function is exported, and I am not planning to remove it. If your goal is to add a border around a single cell, I don't see any issue using it. My advice is to start from a table where all borders are null (`border_remove()`) then use border to add borders on some selected cells.

Does it answer the question?

**Comment by cbailiss** [comment]
Thank you for the response.
I already reviewed the documentation pages before submitting the issue.  As far as I can see, using the functions you list it is not possible for example to set the left border on only the cell that is in the first column, second row, 
e.g. flextable 0.6.6 - vline_left()  does not support the j parameter, e.g.

```
library(officer)
library(flextable)
std_border = fp_border(color="orange")
ft <- flextable(head(iris))
ft <- border_remove(x = ft)
ft <- vline_left(ft, i=2, j=1,  border = std_border)
ft
Error in vline_left(ft, i = 2, j = 1, border = std_border) : 
  unused argument (j = 1)
```

This means the only option is to use border() it seems in this case.

**Comment by cbailiss** [comment]
Actually, the above works without j=1, but then I need to write logic that says if column = 1, use vline_left(), otherwise use vline().  So it can be done, it is just counter-intuitive (to me at least!).

**Comment by davidgohel** [comment]
Could you provide a reproducible example we could use as a base for demonstration?

**Comment by davidgohel** [comment]
I don't think you are using the right function. 

vline_left:

> The function is applying vertical borders to the left side of one or all parts of a flextable. The line is the left border of selected cells of the first column.

**Comment by davidgohel** [comment]
So both are producing the same in HTML:

```
library(officer)
library(flextable)
std_border = fp_border(color="orange")
ft <- flextable(head(iris))
ft <- border_remove(x = ft)
ft <- vline_left(ft, i=2,border = std_border, part = "body")
ft <- vline(ft, i=2, j=1, border = std_border, part = "body")
ft <- hline(ft, i=2, j=1, border = std_border, part = "body")
ft <- hline(ft, i=1, j=1, border = std_border, part = "body")
ft


library(officer)
library(flextable)
std_border = fp_border(color="orange")
ft <- flextable(head(iris))
ft <- border_remove(x = ft)
ft <- border(ft, i=2, j=1, border = std_border, part = "body")
ft
```

As said, if you want to use border, you can use it.

**Comment by davidgohel** [comment]
My view of these functions is:

* `vline`, `hline`, etc. are functions that satisfy 95% of the needs
* `border` should answer all needs, but is not that easy to use because of the overlapping issue and because care is needed when merging cells. I discourage its use for end users.

**Comment by cbailiss** [comment]
Thanks for all of your replies.  I hope the example below helps explain.

**Requirement**

Highlight a specific cell in a table (by setting the background colour and left+right borders - this is part of the branding we use, i.e. only set two borders for cell highlights).
Code should be able to highlight any cell in the body of the table specified by row and column numbers.

Example:

![image](https://user-images.githubusercontent.com/11745789/119474760-6d0e1f80-bd44-11eb-917d-0252ebe0a679.png)

**Approach 1 - using public functions**

```
library(officer)
library(flextable)

# specify cell to highlight
cell_highlight_row_number = 1
cell_highlight_column_number = 1

std_border = fp_border(color="orange")
ft <- flextable(head(iris))
ft <- border_remove(x = ft)

# set background colour
ft <- bg(ft, i=cell_highlight_row_number, j=cell_highlight_column_number, bg="yellow")

# set left and right borders (using public functions)
if (cell_highlight_column_number==1) { 
  ft <- vline_left(ft, i=cell_highlight_row_number, border=std_border, part="body")
} else {
  ft <- vline(ft, i=cell_highlight_row_number, j=cell_highlight_column_number-1, 
              border=std_border, part="body")
}
if (cell_highlight_column_number==5) { 
  ft <- vline_right(ft, i=cell_highlight_row_number, border=std_border, part="body")
} else {
  ft <- vline(ft, i=cell_highlight_row_number, j=cell_highlight_column_number, 
              border=std_border, part="body")
}

# render
ft
```

**Approach 2 - using internal function**

```
library(officer)
library(flextable)

# specify cell to highlight
cell_highlight_row_number = 1
cell_highlight_column_number = 1

std_border = fp_border(color="orange")
ft <- flextable(head(iris))
ft <- border_remove(x = ft)

# set background colour
ft <- bg(ft, i=cell_highlight_row_number, j=cell_highlight_column_number, bg="yellow")

# set left and right borders (using internal function)
ft <- border(ft, i=cell_highlight_row_number, j=cell_highlight_column_number, 
             border.left=std_border, border.right=std_border, part="body")

# render
ft
```

**Comparison**

The code using the internal function is:

- Shorter
- Simpler to read - more descriptive of intent, less logic to have to read through
- border() is more consistent with how the other styling functions like bg() work - i.e. just specify the cell(s) and the styling

I understand your thinking with the public functions however - just wanted to provide some feedback.
And thanks again for a great package.

**Comment by davidgohel** [comment]
Thanks

Let's keep that issue opened until I complete the documentation. I will work on that, it is necessary. 

(one issue with your example can be seen on pdf output), latex or at least flextable implementation does not have the border concept, instead it uses vline or hline (that why also these functions exist in the package also). So the correct code would be:

```
library(officer)
library(flextable)

# specify cell to highlight
cell_highlight_row_number = 1
cell_highlight_column_number = 1

std_border = fp_border(color="orange")
ft <- flextable(head(iris))
ft <- border_remove(x = ft)

# set background colour
ft <- bg(ft, i=cell_highlight_row_number, j=cell_highlight_column_number, bg="yellow")

# set left and right borders (using internal function)
ft <- border(ft, i=cell_highlight_row_number, j=cell_highlight_column_number, 
             border.left=std_border, border.right=std_border, part="body")
ft <- border(ft, i=cell_highlight_row_number, j=cell_highlight_column_number+1, 
             border.left=std_border, part="body")

# render
ft
```


But don't worry about that if you don't want to support PDF. I will try to list the known issues with borders.

**Comment by cbailiss** [comment]
Thanks - and thanks for highlighting the issue for PDF output.  Latex is a whole different universe from my experience working with it.

Would one option be to create a public version of border() that wraps the current internal version?  The public version could then translate the arguments (e.g. border.right) into something that is PDF compatible, e.g. converting it to border.left with column+1.
